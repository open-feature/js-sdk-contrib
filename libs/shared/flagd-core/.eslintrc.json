{
  "extends": ["../../../.eslintrc.json"],
  "ignorePatterns": ["!**/*", "test-harness", "flagd-schemas", "spec"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["*.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/dependency-checks": "error"
      }
    }
  ]
}
