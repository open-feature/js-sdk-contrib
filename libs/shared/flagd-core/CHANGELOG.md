# Changelog

## [0.1.11](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.10...flagd-core-v0.1.11) (2024-02-14)


### 🐛 Bug Fixes

* init in-process error, throw on invalid rules ([#767](https://github.com/open-feature/js-sdk-contrib/issues/767)) ([e9f9e74](https://github.com/open-feature/js-sdk-contrib/commit/e9f9e74d66e9f8666eebb8d06141fce713c7914c))

## [0.1.10](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.9...flagd-core-v0.1.10) (2024-01-30)


### 🐛 Bug Fixes

* falsy boolean shorthand logic ([#746](https://github.com/open-feature/js-sdk-contrib/issues/746)) ([0772c90](https://github.com/open-feature/js-sdk-contrib/commit/0772c90c10906e47109567ba1ac35fe8b38fbe74))

## [0.1.9](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.8...flagd-core-v0.1.9) (2024-01-29)


### 🐛 Bug Fixes

* update disabled behavior to match spec  ([#744](https://github.com/open-feature/js-sdk-contrib/issues/744)) ([3f6b4f4](https://github.com/open-feature/js-sdk-contrib/commit/3f6b4f43e7e79a70517d1d654355cf4b82a31188))

## [0.1.8](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.7...flagd-core-v0.1.8) (2024-01-08)


### 🧹 Chore

* throw ParseError on invalid flagd config ([#714](https://github.com/open-feature/js-sdk-contrib/issues/714)) ([837bf08](https://github.com/open-feature/js-sdk-contrib/commit/837bf0887a8b68e6418963160344af1aaeabbf0a))

## [0.1.7](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.6...flagd-core-v0.1.7) (2023-12-18)


### 🐛 Bug Fixes

* re-add browser support ([#706](https://github.com/open-feature/js-sdk-contrib/issues/706)) ([c262c66](https://github.com/open-feature/js-sdk-contrib/commit/c262c66497e0cc7d8b7ea2d9cc5b85f5d31093e6))

## [0.1.6](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.5...flagd-core-v0.1.6) (2023-12-18)


### ✨ New Features

* **flagd-core:** add update config support, returns changed keys ([#703](https://github.com/open-feature/js-sdk-contrib/issues/703)) ([fa393a6](https://github.com/open-feature/js-sdk-contrib/commit/fa393a6c03ddeacc2899db4d4911cb06712211ba))

## [0.1.5](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.4...flagd-core-v0.1.5) (2023-12-15)


### 🐛 Bug Fixes

* treat empty targeting rules as static ([#696](https://github.com/open-feature/js-sdk-contrib/issues/696)) ([8bff89d](https://github.com/open-feature/js-sdk-contrib/commit/8bff89d023486734a739dbdfb016b2966ec43436))

## [0.1.4](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.3...flagd-core-v0.1.4) (2023-12-13)


### 🐛 Bug Fixes

* fixed issue with nested fractional evaluations ([#686](https://github.com/open-feature/js-sdk-contrib/issues/686)) ([e0dbfdb](https://github.com/open-feature/js-sdk-contrib/commit/e0dbfdb2c73b45bdbadb22b4198b0134e395548a))


### 🧹 Chore

* improve logger, parsing and add helpers ([#689](https://github.com/open-feature/js-sdk-contrib/issues/689)) ([fa0a238](https://github.com/open-feature/js-sdk-contrib/commit/fa0a238bc4533e431e2c2969303866e74f4f181f))

## [0.1.3](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.2...flagd-core-v0.1.3) (2023-12-06)


### 🐛 Bug Fixes

* "in" op by using json-logic-js ([#671](https://github.com/open-feature/js-sdk-contrib/issues/671)) ([4b5e2fe](https://github.com/open-feature/js-sdk-contrib/commit/4b5e2fe5cf89385a8cf5e6be5b4bc0a50d4b791d))
* orphaned grpc connection, semver ~, change events ([#654](https://github.com/open-feature/js-sdk-contrib/issues/654)) ([5afbea7](https://github.com/open-feature/js-sdk-contrib/commit/5afbea754983f95858bf1bdfd15ab51793b0b72e))

## [0.1.2](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.1...flagd-core-v0.1.2) (2023-11-14)


### 🐛 Bug Fixes

* false positive of a falsy variant value check ([#651](https://github.com/open-feature/js-sdk-contrib/issues/651)) ([1c7dc66](https://github.com/open-feature/js-sdk-contrib/commit/1c7dc660d15e00f84ad303d373417f8bb7b71966))


### 🧹 Chore

* address lint issues ([#642](https://github.com/open-feature/js-sdk-contrib/issues/642)) ([bbd9aee](https://github.com/open-feature/js-sdk-contrib/commit/bbd9aee896dc4a0817f379b799a1b8d331ee76c6))

## [0.1.1](https://github.com/open-feature/js-sdk-contrib/compare/flagd-core-v0.1.0...flagd-core-v0.1.1) (2023-11-02)


### ✨ New Features

* flagd-core json logic evaluator ([#623](https://github.com/open-feature/js-sdk-contrib/issues/623)) ([72eacd3](https://github.com/open-feature/js-sdk-contrib/commit/72eacd33ab7147d7348ee125c57282bccd3af9d5))
* Initial version of the flagd js core ([#620](https://github.com/open-feature/js-sdk-contrib/issues/620)) ([8fac8cb](https://github.com/open-feature/js-sdk-contrib/commit/8fac8cb902c8803200b3dbc74eace3d623746b4e))


### 🧹 Chore

* make @openfeature/core peer of flagd/core ([#631](https://github.com/open-feature/js-sdk-contrib/issues/631)) ([3db7c0c](https://github.com/open-feature/js-sdk-contrib/commit/3db7c0c739c84be9fa9cedb87b5e0521a1a0d89c))
