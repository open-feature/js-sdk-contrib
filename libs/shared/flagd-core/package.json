{
  "name": "@openfeature/flagd-core",
  "version": "0.1.11",
  "scripts": {
    "publish-if-not-exists": "cp $NPM_CONFIG_USERCONFIG .npmrc && if [ \"$(npm show $npm_package_name@$npm_package_version version)\" = \"$(npm run current-version -s)\" ]; then echo 'already published, skipping'; else npm publish --access public; fi",
    "current-version": "echo $npm_package_version"
  },
  "peerDependencies": {
    "@openfeature/core": ">=0.0.16"
  },
  "dependencies": {
    "ajv": "^8.12.0",
    "tslib": "^2.3.0"
  }
}
