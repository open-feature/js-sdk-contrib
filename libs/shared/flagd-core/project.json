{
  "name": "flagd-core",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/shared/flagd-core/src",
  "projectType": "library",
  "targets": {
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "npm run publish-if-not-exists",
        "cwd": "dist/libs/shared/flagd-core"
      },
      "dependsOn": ["build"]
    },
    "pullTestHarness": {
      "executor": "nx:run-commands",
      "options": {
        "commands": ["git submodule update --init spec", "git submodule update --init test-harness"],
        "cwd": "libs/shared/flagd-core",
        "parallel": false
      }
    },
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"]
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/shared/flagd-core/jest.config.ts"
      }
    },
    "package": {
      "executor": "@nx/rollup:rollup",
      "outputs": ["{options.outputPath}"],
      "dependsOn": ["pullTestHarness"],
      "options": {
        "project": "libs/shared/flagd-core/package.json",
        "outputPath": "dist/libs/shared/flagd-core",
        "entryFile": "libs/shared/flagd-core/src/index.ts",
        "tsConfig": "libs/shared/flagd-core/tsconfig.lib.json",
        "compiler": "tsc",
        "generateExportsField": true,
        "umdName": "flagd-core",
        "external": "all",
        "format": ["cjs", "esm"],
        "assets": [
          {
            "glob": "package.json",
            "input": "./assets",
            "output": "./src/"
          },
          {
            "glob": "LICENSE",
            "input": "./",
            "output": "./"
          },
          {
            "glob": "README.md",
            "input": "./libs/shared/flagd-core",
            "output": "./"
          }
        ]
      }
    }
  },
  "tags": []
}
