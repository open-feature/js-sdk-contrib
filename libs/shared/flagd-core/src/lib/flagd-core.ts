import {MemoryStorage, Storage} from './storage';
import {
  EvaluationContext,
  FlagNotFoundError,
  FlagValue,
  JsonValue,
  ResolutionDetails,
  StandardResolutionReasons,
  TypeMismatchError,
} from '@openfeature/core';
import {Targeting} from "./targeting/targeting";

/**
 * Expose flag configuration setter and flag resolving methods.
 */
export class FlagdCore {
  private _storage: Storage;
  private _targeting = new Targeting();

  /**
   * Optionally construct with your own storage layer.
   */
  constructor(storage?: Storage) {
    this._storage = storage ? storage : new MemoryStorage();
  }

  /**
   * Add flag configurations to the storage.
   */
  setConfigurations(cfg: string): void {
    this._storage.setConfigurations(cfg);
  }

  resolveBooleanEvaluation(
    flagKey: string,
    defaultValue: boolean,
    evalCtx: EvaluationContext,
  ): ResolutionDetails<boolean> {
    return this.resolve(flagKey, defaultValue, evalCtx, (v) => {
      return typeof v === 'boolean';
    });
  }

  resolveStringEvaluation(
    flagKey: string,
    defaultValue: string,
    evalCtx: EvaluationContext,
  ): ResolutionDetails<string> {
    return this.resolve(flagKey, defaultValue, evalCtx, (v) => {
      return typeof v === 'string';
    });
  }

  resolveNumberEvaluation(
    flagKey: string,
    defaultValue: number,
    evalCtx: EvaluationContext,
  ): ResolutionDetails<number> {
    return this.resolve(flagKey, defaultValue, evalCtx, (v) => {
      return typeof v === 'number';
    });
  }

  resolveObjectEvaluation<T extends JsonValue>(
    flagKey: string,
    defaultValue: T,
    evalCtx: EvaluationContext,
  ): ResolutionDetails<T> {
    return this.resolve(flagKey, defaultValue, evalCtx, (v) => {
      return typeof v === 'object';
    });
  }

  private resolve<T extends FlagValue>(
    flagKey: string,
    defaultValue: T,
    evalCtx: EvaluationContext,
    guard: (a: FlagValue) => boolean,
  ): ResolutionDetails<T> {

    // flag exist check
    const flag = this._storage.getFlag(flagKey);
    if (!flag) {
      throw new FlagNotFoundError(`flag: ${flagKey} not found`);
    }

    // flag status check
    if (flag.state === 'DISABLED') {
      return {
        value: defaultValue,
        reason: StandardResolutionReasons.DISABLED,
      }
    }

    let variant;
    let reason;

    if (!flag.targeting) {
      variant = flag.defaultVariant;
      reason = StandardResolutionReasons.STATIC;
    } else {
      let targetingResolution;
      try {
        targetingResolution = this._targeting.applyTargeting(flagKey, flag.targeting, evalCtx);
      } catch (e) {
        console.error(`Error evaluating targeting rule for flag ${flagKey}, falling back to default`, e);
        targetingResolution = null;
      }

      if (!targetingResolution) {
        variant = flag.defaultVariant;
        reason = StandardResolutionReasons.DEFAULT;
      } else {
        variant = targetingResolution;
        reason = StandardResolutionReasons.TARGETING_MATCH;
      }
    }

    if (typeof variant !== 'string') {
      throw new TypeMismatchError('Variant must be a string, but found ' + typeof variant)
    }

    const resolvedVariant = flag.variants.get(variant)
    if (!resolvedVariant) {
      throw new TypeMismatchError(`Variant ${variant} not found in flag with key ${flagKey}`);
    }

    if (!guard(resolvedVariant)) {
      throw new TypeMismatchError('Evaluated type does not match the flag type');
    }

    return {
      value: resolvedVariant as T,
      reason: reason,
      variant,
    };
  }
}
