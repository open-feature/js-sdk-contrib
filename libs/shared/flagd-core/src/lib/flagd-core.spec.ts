import {FlagdCore} from './flagd-core';
import {StandardResolutionReasons, TypeMismatchError} from '@openfeature/core';

describe('flagd-core resolving', () => {
  const flagCfg = `{"flags":{"myBoolFlag":{"state":"ENABLED","variants":{"on":true,"off":false},"defaultVariant":"on"},"myStringFlag":{"state":"ENABLED","variants":{"key1":"val1","key2":"val2"},"defaultVariant":"key1"},"myFloatFlag":{"state":"ENABLED","variants":{"one":1.23,"two":2.34},"defaultVariant":"one"},"myIntFlag":{"state":"ENABLED","variants":{"one":1,"two":2},"defaultVariant":"one"},"myObjectFlag":{"state":"ENABLED","variants":{"object1":{"key":"val"},"object2":{"key":true}},"defaultVariant":"object1"},"fibAlgo":{"variants":{"recursive":"recursive","memo":"memo","loop":"loop","binet":"binet"},"defaultVariant":"recursive","state":"ENABLED","targeting":{"if":[{"$ref":"emailWithFaas"},"binet",null]}},"targetedFlag":{"variants":{"first":"AAA","second":"BBB","third":"CCC"},"defaultVariant":"first","state":"ENABLED","targeting":{"if":[{"in":["@openfeature.dev",{"var":"email"}]},"second",{"in":["Chrome",{"var":"userAgent"}]},"third",null]}}},"$evaluators":{"emailWithFaas":{"in":["@faas.com",{"var":["email"]}]}}}`;
  let core: FlagdCore;

  beforeAll(() => {
    core = new FlagdCore();
    core.setConfigurations(flagCfg);
  });

  it('should resolve boolean flag', () => {
    const resolved = core.resolveBooleanEvaluation('myBoolFlag', false, {});
    expect(resolved.value).toBeTruthy();
    expect(resolved.reason).toBe(StandardResolutionReasons.STATIC);
    expect(resolved.variant).toBe("on")
  });

  it('should resolve string flag', () => {
    const resolved = core.resolveStringEvaluation('myStringFlag', 'key2', {});
    expect(resolved.value).toBe('val1');
    expect(resolved.reason).toBe(StandardResolutionReasons.STATIC);
    expect(resolved.variant).toBe("key1")
  });

  it('should resolve number flag', () => {
    const resolved = core.resolveNumberEvaluation('myFloatFlag', 2.34, {});
    expect(resolved.value).toBe(1.23);
    expect(resolved.reason).toBe(StandardResolutionReasons.STATIC);
    expect(resolved.variant).toBe("one")
  });

  it('should resolve object flag', () => {
    const resolved = core.resolveObjectEvaluation('myObjectFlag', {key: true}, {});
    expect(resolved.value).toStrictEqual({key: 'val'});
    expect(resolved.reason).toBe(StandardResolutionReasons.STATIC);
    expect(resolved.variant).toBe("object1")
  });

});

describe('flagd-core targeting evaluations', () => {

  const targetingFlag = '{"flags":{"targetedFlag":{"variants":{"first":"AAA","second":"BBB","third":"CCC"},"defaultVariant":"first","state":"ENABLED","targeting":{"if":[{"in":["@openfeature.dev",{"var":"email"}]},"second",null]}}}}';
  let core: FlagdCore;

  beforeAll(() => {
    core = new FlagdCore();
    core.setConfigurations(targetingFlag)
  });

  it('should resolve for correct inputs', () => {
    const resolved = core.resolveStringEvaluation("targetedFlag", "none", {email: "admin@openfeature.dev"});
    expect(resolved.value).toBe("BBB")
    expect(resolved.reason).toBe(StandardResolutionReasons.TARGETING_MATCH)
    expect(resolved.variant).toBe("second")
  });

  it('should fallback to default - missing targeting context data', () => {
    const resolved = core.resolveStringEvaluation("targetedFlag", "none", {});
    expect(resolved.value).toBe("AAA")
    expect(resolved.reason).toBe(StandardResolutionReasons.DEFAULT)
    expect(resolved.variant).toBe("first")
  });

});

describe('flagd-core validations', () => {
  // flags of disabled, invalid variants and missing variant
  const mixFlags =
    '{"flags":{"myBoolFlag":{"state":"DISABLED","variants":{"on":true,"off":false},"defaultVariant":"on"},"myStringFlag":{"state":"ENABLED","variants":{"on":true,"off":false},"defaultVariant":"on"},"myIntFlag":{"state":"ENABLED","variants":{"two":2},"defaultVariant":"one"}}}';
  let core: FlagdCore;

  beforeAll(() => {
    core = new FlagdCore();
    core.setConfigurations(mixFlags);
  });

  it('should validate flag type - eval int as boolean', () => {
    expect(() => core.resolveBooleanEvaluation('myIntFlag', true, {}))
      .toThrow(TypeMismatchError)
  });

  it('should validate flag status', () => {
    const evaluation = core.resolveBooleanEvaluation('myBoolFlag', false, {});

    expect(evaluation).toBeTruthy()
    expect(evaluation.value).toBe(false)
    expect(evaluation.reason).toBe(StandardResolutionReasons.DISABLED)
  });

  it('should validate variant', () => {
    expect(() => core.resolveStringEvaluation('myStringFlag', 'hello', {}))
      .toThrow(TypeMismatchError)
  });

  it('should validate variant existence', () => {
    expect(() => core.resolveNumberEvaluation('myIntFlag', 100, {}))
      .toThrow(TypeMismatchError)
  });
});
