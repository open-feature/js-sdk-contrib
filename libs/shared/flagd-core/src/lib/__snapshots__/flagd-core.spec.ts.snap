// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flagd-core resolving falsy variant values should resolve all flags 1`] = `
[
  {
    "flagKey": "myBoolFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": false,
    "variant": "off",
  },
  {
    "flagKey": "myStringFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": "",
    "variant": "key1",
  },
  {
    "flagKey": "myFloatFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": 0,
    "variant": "zero",
  },
  {
    "flagKey": "myIntFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": 0,
    "variant": "zero",
  },
  {
    "flagKey": "myObjectFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": {},
    "variant": "object1",
  },
  {
    "flagKey": "fibAlgo",
    "flagMetadata": {},
    "reason": "DEFAULT",
    "value": "recursive",
    "variant": "recursive",
  },
  {
    "flagKey": "targetedFlag",
    "flagMetadata": {},
    "reason": "DEFAULT",
    "value": "AAA",
    "variant": "first",
  },
]
`;

exports[`flagd-core resolving truthy variant values should resolve all flags 1`] = `
[
  {
    "flagKey": "myBoolFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": true,
    "variant": "on",
  },
  {
    "flagKey": "myStringFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": "val1",
    "variant": "key1",
  },
  {
    "flagKey": "myFloatFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": 1.23,
    "variant": "one",
  },
  {
    "flagKey": "myIntFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": 1,
    "variant": "one",
  },
  {
    "flagKey": "myObjectFlag",
    "flagMetadata": {},
    "reason": "STATIC",
    "value": {
      "key": "val",
    },
    "variant": "object1",
  },
  {
    "flagKey": "fibAlgo",
    "flagMetadata": {},
    "reason": "DEFAULT",
    "value": "recursive",
    "variant": "recursive",
  },
  {
    "flagKey": "targetedFlag",
    "flagMetadata": {},
    "reason": "DEFAULT",
    "value": "AAA",
    "variant": "first",
  },
]
`;
